### PR TITLE
Add handling for 'time' variable to WindowedZarrLoader

### PR DESCRIPTION
**Pull request by mcgibbon**
This PR adds the ability to load time as a number of seconds since epoch to the WindowedZarrLoader.

Refactored public API:
- If "time" is requested from the WindowedZarrLoader, it will be returned in records as a number of samples since epoch (since 1970-01-01)
- SyntheticZarrLoader can now include time as a record variable
- WindowedZarrLoader now returns variables in the dimension order they were requested. This behavior was unintentionally changed in an earlier commit.
- WindowedZarrLoader can now return variables which are missing any of the requested dimensions, not just the last dimension. This was required in order to return the "time" variable.

- [x] Tests added

Coverage reports (updated automatically):

### CODE REVIEW

File: external/fv3fit/fv3fit/data/tfdataset.py [hunk 1]
```diff
@@ -16,8 +17,7 @@
 
 def stack(ds: xr.Dataset, unstacked_dims: Sequence[str]):
     stack_dims = [dim for dim in ds.dims if dim not in unstacked_dims]
-    unstacked_dims = [dim for dim in ds.dims if dim in unstacked_dims]
-    unstacked_dims.sort()  # needed to always get [x, y, z] dimensions
```
Comment (oliverwm1) [inline external/fv3fit/fv3fit/data/tfdataset.py:20]: was this `.sort()` causing an issue for your feature? Or did you just remove since it's unnecessary?

File: external/fv3fit/fv3fit/data/tfdataset.py [hunk 2]
```diff
@@ -119,6 +118,10 @@ class WindowedZarrLoader(TFDatasetLoader):
     necessarily appear in each iteration over the tf.data.Dataset. Each sample window
     is independently selected along any stacked (sample) dimensions.
 
+    If the "time" variable itself is loaded, it will be converted to a number of
+    seconds since 1970-01-01. It is also required that the time variable be
+    one-dimensional in "time" (as opposed to varying along some perturbation axis).
```
Comment (oliverwm1) [inline external/fv3fit/fv3fit/data/tfdataset.py:123]: This conversion is pretty specific for the general purpose `WindowedZarrLoader`. (looks like the class mostly just does subselection right now, no conversion of actual values?)

File: external/fv3fit/fv3fit/data/tfdataset.py [hunk 3]
```diff
@@ -47,21 +47,20 @@ def __post_init__(self):
             raise TypeError("times must be one of 'window' or 'start'")
 
     def get_record(self, name: str, ds: xr.Dataset, unstacked_dims: Sequence[str]):
-        for dim in unstacked_dims[:-1]:
-            if dim not in ds[name].dims:
-                raise ValueError("variable {} has no dimension {}".format(name, dim))
         if self.times == "start":
             ds = ds.isel(time=0)
-        ds = stack(ds, unstacked_dims)
+        # must subset dataset to one variable to avoid attaching dimensions
```
Comment (brianhenn) [inline external/fv3fit/fv3fit/data/tfdataset.py:52]: Couldn't we just make `stack` operate on data arrays instead and not have to worry about this?

File: external/fv3fit/fv3fit/data/tfdataset.py (hunk 2 above)
Comment (brianhenn) [reply external/fv3fit/fv3fit/data/tfdataset.py:123]: Would tend to agree with Oli, could the time conversion be moved into the CycleGAN training function? I know the CycleGAN is the only user of time from the `WindowedZarrLoader` at the moment but maybe it won't be.

File: external/fv3fit/fv3fit/data/tfdataset.py (hunk 1 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/data/tfdataset.py:20]: It was causing issues. The tests don't care, but it broke my models for production because my post-evaluation scripts expect specific dimension orders, and they were no longer followed when I added `.sort` in a previous commit.

File: external/fv3fit/fv3fit/data/tfdataset.py (hunk 2 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/data/tfdataset.py:123]: It is and it isn't - the time variable has to be converted to a data type tensorflow can parse (there's no such thing as a cftime.DatetimeJulian tensor). This was the most unopinionated conversion I could think of.

File: external/fv3fit/fv3fit/data/tfdataset.py (hunk 1 above)
Comment (brianhenn) [reply external/fv3fit/fv3fit/data/tfdataset.py:20]: Seems fine, but can you jog my memory why you had commented in before that a specific order was needed?

File: external/fv3fit/fv3fit/data/tfdataset.py (hunk 3 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/data/tfdataset.py:52]: We probably could. It may even "just work" - I'll try.

File: external/fv3fit/fv3fit/data/tfdataset.py (hunk 1 above)
Comment (oliverwm1) [reply external/fv3fit/fv3fit/data/tfdataset.py:20]: Yes... this code is used is many places, so some care required here IMO.

File: external/fv3fit/fv3fit/data/tfdataset.py (hunk 2 above)
Comment (brianhenn) [reply external/fv3fit/fv3fit/data/tfdataset.py:123]: OK thanks, I wasn't connecting that time could never be in a tf dataset as a datetime object, I'm fine keeping it then. Maybe just clean up the `records` function a bit by moving the specific time handling into its own function, and think about whether this could just be tested via a functional unit test instead of (in addition to?) a test on the loader.

File: external/fv3fit/fv3fit/data/tfdataset.py (hunk 3 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/data/tfdataset.py]: It needed some other changes, but they're for the better. Updated.

File: external/fv3fit/fv3fit/data/tfdataset.py (hunk 2 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/data/tfdataset.py:124]: I think the code itself for the time handling within `generator` is simple enough that it doesn't need a test - it's the same datetime conversion code people copy paste from stack overflow all the time. When I added the loader test, it caught a fair number of failures to do with making sure the result is aligned correctly with the other data and has the right dimensions, and to do with things I hadn't thought of when loading different types of zarr data (e.g. the data type is different if you save cftime vs datetime). So I think the larger-scale test is needed.

It's actually quite nice having tests at the level of the loader because they're meant to be fairly backwards-compatible and really thought of as a "unit", so changes to implementation details rarely require the tests themselves to be updated/refactored, only added to.

**Review by brianhenn — APPROVED** [review]
Thanks @mcgibbon. Approving with a couple requests for continued cleanup of the stacking routine.

File: external/fv3fit/fv3fit/data/tfdataset.py [hunk 4]
```diff
@@ -49,10 +49,11 @@ def __post_init__(self):
     def get_record(self, name: str, ds: xr.Dataset, unstacked_dims: Sequence[str]):
         if self.times == "start":
             ds = ds.isel(time=0)
-        # must subset dataset to one variable to avoid attaching dimensions
+        # must stack array instead of dataset dataset to avoid attaching dimensions
         # to it that happen to exist in other variables
-        ds = stack(ds[[name]], unstacked_dims)
-        return ds[name].values
+        array = stack(ds[name], unstacked_dims)
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/data/tfdataset.py:54]: Good suggestion, updated.

File: external/fv3fit/fv3fit/data/tfdataset.py [hunk 5]
```diff
@@ -49,10 +49,11 @@ def __post_init__(self):
     def get_record(self, name: str, ds: xr.Dataset, unstacked_dims: Sequence[str]):
         if self.times == "start":
             ds = ds.isel(time=0)
-        # must subset dataset to one variable to avoid attaching dimensions
+        # must stack array instead of dataset dataset to avoid attaching dimensions
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/data/tfdataset.py:52]: With your suggested updates this is a little more obvious. Removed.
